src/VirtoCommerce.Platform.Caching/Redis/RedisPlatformMemoryCache.cs

```diff
@@ -1,47 +1,90 @@
+        private readonly IMemoryCache _memoryCache;
+        private readonly CachingOptions _cachingOptions;
         private readonly RedisCachingOptions _redisCachingOptions;
         private readonly ILogger _log;
-        private readonly RetryPolicy _retryPolicy;
+        private readonly TelemetryClient _telemetryClient;
 
         private static readonly string _cacheId = Guid.NewGuid().ToString("N");
         
 
         public RedisPlatformMemoryCache(IMemoryCache memoryCache, IOptions<CachingOptions> options
+            , ConnectionMultiplexer connection
             , ISubscriber bus
+            , IOptions<CachingOptions> cachingOptions
             , IOptions<RedisCachingOptions> redisCachingOptions
             , ILogger<RedisPlatformMemoryCache> log
+            , TelemetryClient telemetryClient
             ) : base(memoryCache, options, log)
        {
+            _memoryCache = memoryCache;
             _log = log;
+            _telemetryClient = telemetryClient;
             _bus = bus;
 
+            _cachingOptions = cachingOptions.Value;
             _redisCachingOptions = redisCachingOptions.Value;
-            _bus.Unsubscribe(_redisCachingOptions.ChannelName);
-            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage);
+
+            connection.ConnectionFailed += OnConnectionFailed;
+            connection.ConnectionRestored += OnConnectionRestored;
+
+            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage, CommandFlags.FireAndForget);
 
             _log.LogInformation($"{nameof(RedisPlatformMemoryCache)}: subscribe to channel {_redisCachingOptions.ChannelName } current instance:{ _cacheId }");
+            _telemetryClient.TrackEvent("RedisSubscribed", new Dictionary<string, string>
+            {
+                {"channelName", _redisCachingOptions.ChannelName},
+                {"cacheId", _cacheId}
+            });
+        }
+
+        protected virtual void OnConnectionFailed(object? sender, ConnectionFailedEventArgs e)
+        {
+            _log.LogError($"Redis disconnected from instance { _cacheId }. Endpoint is {e.EndPoint}, failure type is {e.FailureType}");
+            _telemetryClient.TrackException(e.Exception);
+            _telemetryClient.TrackEvent("RedisDisconnected", new Dictionary<string, string>
+            {
+                {"channelName", _redisCachingOptions.ChannelName},
+                {"cacheId", _cacheId},
+                {"endpoint", e.EndPoint.ToString()},
+                {"failureType", e.FailureType.ToString()}
+            });
 
-            _retryPolicy = Policy.Handle<Exception>().WaitAndRetry(
-                _redisCachingOptions.BusRetryCount, retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt - 1)));
+            CacheEnabled = false;
+            GlobalCacheRegion.ExpireRegion();
+        }
+
+        protected virtual void OnConnectionRestored(object? sender, ConnectionFailedEventArgs e)
```

Comment:
What is the goal to do this and other methods as virtual and protected? Is there is a small chance of overriding in the spectral future? :)
We have an important logic in methods, these logics should be tested by mocks verification, eg:
in method OnConnectionFailed we have cache disabling, this important part can be tested by unit-testing.
Also, we can verify the methods invocations eg: TrackException, TrackEvent

Reply:
In platform it's always better to make methods at least protected and virtual for extensibility. You're yourself saying it has important logic. For some reason someone may be want to override this logic,

I'll add unit tests, because of this that PR is draft currently

Reply:
I think methods with these modifiers will be never overridden, cuz we have ServiceCollectionExtensions with the only single implementation, and it is a quite general platform thing. If we want to generalize this, we have to think about pushing up these methods higher. In this implementation, the chance of reusing/overriding is quite small :)